### ISSUE TITLE
How are matching cost computation and horizontal path aggregation fused

### ISSUE THREAD

**Issue by angli66** [dhernandez0/sgm#40]
Dear @dhernandez0 ,

Thank you for this great implementation of SGM. I'm relatively new to CUDA and get a bit confused while studying your code. As mentioned in the paper, the computation of the initial matching cost is fused with horizontal path aggregation. In the code, I can see that the matching cost is done on stream 1, and the leftToRight and rightToLeft computations are done on stream2 and stream3 respectively, which means they are done in parallel. I'm wondering how is this possible? Shouldn't the matching cost be completed before the path aggregation computation starts as the latter depends on the former? Or if there exists some sort of synchronization between these two parts in the code that I didn't discover?

Sorry for disturbing you with this trivial question. I would appreciate it a lot if you can help me understand it.

**Comment by dhernandez0** [comment]
Dear @angli66 

Thanks for your interest in our work. Both, leftToRight and rightToLeft recompute the matching cost at the same time. That's why they can be done in parallel. If you follow the code you'll see they use a flag "recompute=true". Feel free to ask any other questions.

**Comment by angli66** [comment]
Hi,
Thank you for your response! Now I understand the horizontal path is recomputed. When I tested through the code there was another point I noticed. If I turn off "recompute=true" for horizontal path aggregation and synchronize before starting to aggregate, the result matches the original result numerically. However if I turned on recompute for "CostAggregationKernelUpToDown" (the one before the last one), the result seems to have minor value differences from the original result. I'm wondering is there any reason that this is happening or it's simply that "recompute" for vertical path just shouldn't be turned on? Thank you so much!
